File: clients/typescript/src/topic_manager.ts
import { TopicMessage } from './dtos';

export interface IConnectionManager {
    readonly isConnected: boolean;
    send(method: string, ...args: any[]): Promise<void>;
    subscribeTopic(topicInstance: ITopic): void;
    unsubscribeTopic(topicInstance: ITopic): void;
}

export interface ITopic {
    readonly name: string;
    dispatchMessage(message: TopicMessage<unknown>): void;
}

export class TopicManager {
    private readonly connectionManager: IConnectionManager;
    public readonly topicMap: Map<string, Set<ITopic>> = new Map();

    constructor(connectionManager: IConnectionManager) {
        this.connectionManager = connectionManager;
    }

    public subscribeTopic(topicInstance: ITopic): void {
        const topicName = topicInstance.name;
        if (!topicName)
            return;

        if (!this.topicMap.has(topicName))
            this.topicMap.set(topicName, new Set());

        const topicSet = this.topicMap.get(topicName)!;
        const isTopicEmpty = topicSet.size === 0;
        topicSet.add(topicInstance);

        if (isTopicEmpty && this.connectionManager.isConnected)
            this.connectionManager.send('Topic.Subscribe', topicName);
    }

    public unsubscribeTopic(topicInstance: ITopic): void {
        const topicName = topicInstance.name;
        if (!topicName || !this.topicMap.has(topicName))
            return;

        const topicSet = this.topicMap.get(topicName)!;
        topicSet.delete(topicInstance);

        if (topicSet.size === 0) {
            this.topicMap.delete(topicName);
            if (this.connectionManager.isConnected)
                this.connectionManager.send('Topic.Unsubscribe', topicName);
        }
    }

    public resubscribeAll(): void {
        if (!this.connectionManager.isConnected)
            return;

        for (const [topicName, topicSet] of this.topicMap.entries()) {
            if (topicSet.size > 0)
                this.connectionManager.send('Topic.Subscribe', topicName);
        }
    }

    public dispatchMessage(topicName: string, topicMessage: TopicMessage<unknown>): void {
        const topicSet = this.topicMap.get(topicName);
        if (!topicSet)
            return;

        for (const topicInstance of Array.from(topicSet))
            topicInstance.dispatchMessage(topicMessage);
    }
}

File: clients/csharp/AmenoLink/TopicManager.cs
namespace AmenoLink;

internal class TopicManager(IConnectionManager connectionManager)
{
    private readonly Dictionary<string, HashSet<ITopicSubscriber>> topicMap = [];
    private readonly Lock lockObject = new();

    public void SubscribeTopic(ITopicSubscriber topicInstance)
    {
        string topicName = topicInstance.Name;
        if (string.IsNullOrEmpty(topicName))
            return;

        bool isTopicEmpty = false;
        lock (lockObject)
        {
            if (!topicMap.TryGetValue(topicName, out var topicSet))
            {
                topicSet = [];
                topicMap[topicName] = topicSet;
            }

            isTopicEmpty = topicSet.Count == 0;
            topicSet.Add(topicInstance);
        }

        if (isTopicEmpty && connectionManager.IsConnected)
            _ = connectionManager.Send("Topic.Subscribe", topicName);
    }

    public void UnsubscribeTopic(ITopicSubscriber topicInstance)
    {
        string topicName = topicInstance.Name;
        if (string.IsNullOrEmpty(topicName))
            return;

        bool shouldUnsubscribeOnServer = false;
        lock (lockObject)
        {
            if (!topicMap.TryGetValue(topicName, out var topicSet))
                return;

            topicSet.Remove(topicInstance);

            if (topicSet.Count == 0)
            {
                topicMap.Remove(topicName);
                shouldUnsubscribeOnServer = true;
            }
        }

        if (shouldUnsubscribeOnServer && connectionManager.IsConnected)
            _ = connectionManager.Send("Topic.Unsubscribe", topicName);
    }

    public void ResubscribeAll()
    {
        if (!connectionManager.IsConnected)
            return;

        List<string> topicsToResubscribe;
        lock (lockObject)
        {
            topicsToResubscribe = topicMap.Where(kvp => kvp.Value.Count > 0).Select(kvp => kvp.Key).ToList();
        }

        foreach (string topicName in topicsToResubscribe)
            _ = connectionManager.Send("Topic.Subscribe", topicName);
    }

    public void DispatchMessage(string topicName, TopicMessage<object> topicMessage)
    {
        List<ITopicSubscriber> targets;
        lock (lockObject)
        {
            if (!topicMap.TryGetValue(topicName, out var topicSet))
                return;

            targets = [.. topicSet];
        }

        foreach (var topicInstance in targets)
            topicInstance.DispatchMessage(topicMessage);
    }
}
